analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs

```diff
@@ -24,48 +24,78 @@ namespace SonarAnalyzer.Rules.CSharp
                     }
                 }
-            },
-            SyntaxKind.ClassDeclaration,
-            SyntaxKindEx.RecordClassDeclaration);
+            });
 
-        private static bool HasVirtualMembers(TypeDeclarationSyntax typeDeclaration)
-        {
-            var classMembers = typeDeclaration.Members;
-            return classMembers.OfType<MethodDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword))
-                   || classMembers.OfType<PropertyDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword))
-                   || classMembers.OfType<IndexerDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword))
-                   || classMembers.OfType<EventDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword));
-        }
+        private static bool HasVirtualMembers(TypeDeclarationSyntax typeDeclaration) =>
+            typeDeclaration.Members
+                .Where(member => member.IsAnyKind(PossiblyVirtualKinds))
+                .Any(member => member.Modifiers().Any(SyntaxKind.VirtualKeyword));
 
-        private static bool IsPrivateButNotSealedType(TypeDeclarationSyntax typeDeclaration) =>
-            typeDeclaration.Modifiers.Any(SyntaxKind.PrivateKeyword)
-            && !typeDeclaration.Modifiers.Any(SyntaxKind.StaticKeyword)
+        private static bool IsNotSealed(TypeDeclarationSyntax typeDeclaration) =>
+            !typeDeclaration.Modifiers.Any(SyntaxKind.StaticKeyword)
             && !typeDeclaration.Modifiers.Any(SyntaxKind.SealedKeyword)
             && !typeDeclaration.Modifiers.Any(SyntaxKind.AbstractKeyword);
 
-        private static bool IsPrivateTypeInherited(INamedTypeSymbol privateTypeInfo) =>
-            privateTypeInfo.ContainingType.GetAllNamedTypes()
-                                          .Any(symbol => !symbol.MetadataName.Equals(privateTypeInfo.MetadataName)
-                                                         && symbol.DerivesFrom(privateTypeInfo));
+        private static bool IsPossiblyDerived(TypeDeclarationSyntax declaration, Lazy<SemanticModel> model, Lazy<List<INamedTypeSymbol>> otherSymbols, out string modifierDescription)
+        {
+            modifierDescription = string.Empty;
+            if (declaration.Modifiers.Any(SyntaxKind.PrivateKeyword))
+            {
+                modifierDescription = "Private";
+                var symbol = model.Value.GetDeclaredSymbol(declaration);
+                return symbol.ContainingType.GetAllNamedTypes()
+                    .Any(other => !other.MetadataName.Equals(symbol.MetadataName) && other.DerivesFrom(symbol));
+            }
+            if (declaration.Modifiers.Any(SyntaxKindEx.FileKeyword))
+            {
+                modifierDescription = "File-scoped";
+                var symbol = model.Value.GetDeclaredSymbol(declaration);
+                return otherSymbols.Value
```

Comment:
This is a nitpick, I think it's more readable this way but I'll let you decide (it can be applied also on line 87)
```suggestion
                return otherSymbols.Value.Exists(x => 
                	!x.MetadataName.Equals(symbol.MetadataName) 
                	&& x.DerivesFrom(symbol));
```

Reply:
I will ident them as you suggested, I like it as well. I think I will leave the name to be `other` because it feels a bit more readable to me

Reply:
Yeah, I was suggesting `x` based on what I read on [coding-style](https://github.com/SonarSource/sonar-dotnet/blob/master/docs/coding-style.md#naming-conventions). But I guess we can make exceptions if they provide values